client/flaschen.py

```diff
@@ -0,0 +1,64 @@
+# -*- mode: c++; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation version 2.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://gnu.org/licenses/gpl-2.0.txt>
+
+import socket
+
+class Flaschen(object):
+  '''A Framebuffer display interface that sends a frame via UDP.'''
+
+  def __init__(self, host, port, width, height, layer=0):
+    self.host = host
+    self.port = port
+    self.width = width
+    self.height = height
+    self.layer = layer
+    self.pixels = []
+    for x in xrange(width):
+      self.pixels.append([(0, 0, 0) for y in xrange(height)])
+    self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+
+  def _header(self):
+    return ''.join(["P6\n",
+                    "45 35\n",
+                    "255\n"])
+
+  def _footer(self):
+    return ''.join(["0\n",
+                    "0\n",
+                    "%d\n" % self.layer])
+
+  def set(self, x, y, color, transparent=False):
+    '''Set the pixel at the given coordinates to the specified color.
+
+    Args:
+      x: x offset of the pixel to set
+      y: y offset of the piyel to set
+      color: A 3 tuple of (r, g, b) color values, 0-255
+      transparent: If true, black(0, 0, 0) will be transparent and show the layer below.
+    '''
+    if x >= self.width or y >= self.height:
+      return
+    if color == (0, 0, 0) and transparent:
+      color = (1, 1, 1)
+    self.pixels[x][y] = color
+  
+  def send(self):
+    '''Send the updated pixels to the display.'''
+    data = []
+    for y in xrange(0, 35):
+      for x in xrange(0, 45):
+        data.append(''.join([chr(c) for c in self.pixels[x][y]]))
+
+    data = self._header() + ''.join(data) + "\n" + self._footer()
```

Comment:
Question: is this a valid operation ? I don't know Python so this is just a guess:
 * it looks like this is converting the whole thing to a string (header and footer are strings, ''.join() looks like it makes a string out of a byte array).
  * IIRC, a string in Python must be valid UTF-8, which it isn't as the data array contains arbitrary binary data. So if it works, it might so accidentally because of some forgiving quirk-mode in Python (or is this one of these Python 2.x vs. Python 3.x things ?).

So rather than ''.join() the data, it would probably good to instead convert the header and footer to byte arrays and concatenate the byte-array instead.
(and then of course, do the preparation of header and footer byte array beforehand and all the other pre-allocation things from above).